### ISSUE TITLE
support new checker: UseObjectsNonNull/UseObjectsIsNull

### ISSUE THREAD

**Issue by valuetodays** [checkstyle/checkstyle#11235]
In my early life as a java beginning, I sometimes mis-use ==/!=, such like the following code.

```
User user = getFromDb(xxx);
if (user == null) {
}
if (user != null) {
}
```

Typing `if (user == null)`  to `if (user = null)`, which makes error logic.

Using `if (null == user) ` will be better, if you type `if (user = null) `, comiler will tell.

In Java7, Objects.isNull() or Objects.nonNull() is useful.

In this situation, I think we should use  Objects.isNull(user) in stead of `if (user == null)`. 

In my team, we use linux shell script to check code to avoid xxx == null  / xxx != null / null == xxx / null != xxx.

Will checkstyle support a new check such like UseObjectsNonNull/UseObjectsIsNull.

**Comment by nmancus1** [comment]
This can be checked by[ MatchXPath](https://checkstyle.sourceforge.io/config_coding.html#MatchXpath).

Example:

```
➜  src cat Test.java 
public class Test {
    class User {

    }
    void method(User user) {
        if (user == null) { // violation

        }

        if (null == user) { // violation

        }

        if (user != null) { // violation

        }

        if (null != user) { // violation

        }
    }
}
➜  src cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="MatchXpath">
            <property name="query" value="//EXPR/EQUAL/LITERAL_NULL"/>
            <message key="matchxpath.match"
                     value="Use 'Objects.isNull()' instead of null check."/>
        </module>
        <module name="MatchXpath">
            <property name="query" value="//EXPR/NOT_EQUAL/LITERAL_NULL"/>
            <message key="matchxpath.match"
                     value="Use 'Objects.nonNull()' instead of null check."/>
        </module>
    </module>
</module>
➜  src java -jar checkstyle-9.2.1-all.jar -c config.xml Test.java
Starting audit...
[ERROR] Test.java:6:21: Use Objects.isNull() instead of null check. [MatchXpath]
[ERROR] Test.java:10:13: Use Objects.isNull() instead of null check. [MatchXpath]
[ERROR] Test.java:14:21: Use Objects.nonNull() instead of null check. [MatchXpath]
[ERROR Test.java:18:13: Use Objects.nonNull() instead of null check. [MatchXpath]
Audit done.
Checkstyle ends with 4 errors.


```

**Comment by nmancus1** [comment]
I wonder how many teams follow this style... from the Javadoc for [Objects#nonNull](https://docs.oracle.com/javase/8/docs/api/java/util/Objects.html#nonNull-java.lang.Object-): `This method exists to be used as a Predicate, filter(Objects::nonNull)`. 

It seems like this method was introduced at the same time as many other functional programming APIs (to complement streams, etc.), so I would think that using these methods as you've shown above is likely not a very common usage.

Please let us know if the given `MatchXPath` modules above work for your team.

**Comment by valuetodays** [comment]
> I wonder how many teams follow this style... from the Javadoc for [Objects#nonNull](https://docs.oracle.com/javase/8/docs/api/java/util/Objects.html#nonNull-java.lang.Object-): `This method exists to be used as a Predicate, filter(Objects::nonNull)`.
> 
> It seems like this method was introduced at the same time as many other functional programming APIs (to complement streams, etc.), so I would think that using these methods as you've shown above is likely not a very common usage.
> 
> Please let us know if the given `MatchXPath` modules above work for your team.

Yes, some teams use Objects like `filter(Objects::nonNull)`. And the javadoc of Objects is 

```
This class consists of static utility methods for operating
on objects.  These utilities include null-safe or null-tolerant methods for computing the hash code of an object,
returning a string for an object, and comparing two objects.
```
.
Before we replace null != xxx or xxx != null with Objects.nonNull, we check the javadoc, and agree to use this style to avoid `if (xxx = null) ` ,  which should be `if (xxx = null) `. 

It is just a team convention.

**Comment by valuetodays** [comment]
MatchXpath was introduced Since Checkstyle 8.39. 
The version we are current using is 8.29. I will read the release log and upgrade to latest release. Then I will be back to response  again.

**Comment by strkkk** [comment]
note - instead of linux script for grepping source code you can use checkstyle https://checkstyle.org/config_regexp.html#RegexpSinglelineJava or https://checkstyle.org/config_regexp.html#RegexpSingleline with format like `null =` or `null !=`

https://checkstyle.org/config_coding.html#InnerAssignment can be useful as well

**Comment by valuetodays** [comment]
We decide to upgrade checkstyle version from 8.29 to 8.41. 

`MatchXPath`  works for our team. thank @nmancus1 

InnerAssignment is very useful. thanks @strkkk

**Comment by vlsi** [comment]
For reference `Objects.isNull` should not be used instead of regular `null` checks.
Note the documentation:
> https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Objects.html#isNull(java.lang.Object)
>This method exists to be used as a [Predicate](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/function/Predicate.html), filter(Objects::isNull)

In other words, the following code is **good**

```java
if (user == null) {
}
if (user != null) {
}
```

while the following code is **bad**, because it is more obscure without real gain:

```java
if (Objects.isNull(user)) {
}
```

**Comment by vlsi** [comment]
On top of that, it would be better to use tools like OpenRewrite to automatically fix code: https://docs.openrewrite.org/reference/recipes/java/removeobjectsisnull
